### PR TITLE
added multi-threading processing for topic messages; see #7443

### PR DESCRIPTION
**Pull request by dsukhoroslov**
The second try, instead of #7472. Most changes are around TopicConfig now. Please have a look.

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/topic/impl/TopicService.java [hunk 1]
```diff
@@ -145,7 +147,14 @@ public void incrementReceivedMessages(String topicName) {
 
     public void publishEvent(String name, TopicEvent event) {
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/topic/impl/TopicService.java]: What about this one? (by making `counter` an `AtomicInteger`)
- no need to create Long
- no need to find TopicConfig in every run

```
 public void publishEvent(String name, TopicEvent event, boolean multiThreadingEnabled) {
        Collection<EventRegistration> registrations = eventService.getRegistrations(TopicService.SERVICE_NAME, name);
        int orderKey = multiThreadingEnabled ? counter.incrementAndGet() : name.hashCode();
        eventService.publishEvent(TopicService.SERVICE_NAME, registrations, event, orderKey);

eventService.publishEvent(TopicService.SERVICE_NAME, registrations, event, multiThreadingEnabled ? orderKey)
    }
```

Comment (dsukhoroslov) [inline hazelcast/src/main/java/com/hazelcast/topic/impl/TopicService.java]: Yes, looks reasonable. Will do.

Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/topic/impl/TopicService.java]: You can put [here](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/topic/impl/TopicProxySupport.java#L40) a field like `private final TopicConfig config` and can pass `config.multiThreadingEnabled ` to [this method](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/topic/impl/TopicProxySupport.java#L94)

**Comment by dsukhoroslov** [comment]
@ahmetmircik, is it ok now? what else shoud be done in scope of this PR?

File: hazelcast/src/test/java/com/hazelcast/topic/TopicTest.java [hunk 2]
```diff
@@ -627,4 +629,76 @@ public void onMessage(Message msg) {
         assertEquals(1000, stats.getPublishOperationCount());
         assertEquals(2000, stats.getReceiveOperationCount());
     }
+
+    @Test
+    @SuppressWarnings("unchecked")
+    public void testTopicMultiThreading() throws Exception {
```
Comment (ahmetmircik) [inline hazelcast/src/test/java/com/hazelcast/topic/TopicTest.java]: Seems a long running test (longer than 10 seconds) please annotate it with `NightlyTest ` as in [here](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/test/java/com/hazelcast/map/impl/InterceptorRegistryTest.java#L67)

File: hazelcast/src/test/java/com/hazelcast/topic/TopicStressTest.java [hunk 3]
```diff
@@ -56,6 +56,7 @@
     public void setUp() {
         TopicConfig topicConfig = new TopicConfig();
         topicConfig.setName("topic*");
+        topicConfig.setMultiThreadingEnabled(true);
```
Comment (ahmetmircik) [inline hazelcast/src/test/java/com/hazelcast/topic/TopicStressTest.java]: You can do this test parameterized on `multiThreadingEnabled` by setting it to `true/false` in each run as in [here](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/test/java/com/hazelcast/map/nearcache/NearCacheTest.java#L82)

File: hazelcast/src/test/java/com/hazelcast/topic/TopicTest.java (hunk 2 above)
Comment (Donnerbart) [inline hazelcast/src/test/java/com/hazelcast/topic/TopicTest.java]: Just a note: If that method if important for code coverage, please use `SlowTest`. If it's just for correctness or a kind of integration/stress test use `NightlyTest`. All categories will be executed, but just `SlowTest` is considered for our code coverage reports.

**Comment by ahmetmircik** [comment]
@dsukhoroslov i added my last comments when those are fixed i am ok with this PR, thanks for the contribution.

File: hazelcast/src/test/java/com/hazelcast/topic/TopicTest.java (hunk 2 above)
Comment (dsukhoroslov) [inline hazelcast/src/test/java/com/hazelcast/topic/TopicTest.java]: this depends on the machine where tests are runing. on my machine it states
`Finished Running Test: testTopicMultiThreading in 8,576 seconds.`
the test code is the same as in `testTopicLocalOrder` and `testTopicGlobalOrder` and takes similair time. But these tests are not anotated.

**Comment by dsukhoroslov** [comment]
done, please have a look.

**Comment by ahmetmircik** [comment]
seems mergeable to me, guys @eminn @jerrinot can you also have a look please?

**Comment by eminn** [comment]
looks good to merge :+1: 

@dsukhoroslov thanks for the contribution.

**Comment by ahmetmircik** [comment]
run-lab-run

**Comment by ahmetmircik** [comment]
@dsukhoroslov can you please fix check-style errors in this PR? 
To see the errors, you can run the checks via : `mvn verify -Pcheckstyle -DskipTests`

**Comment by dsukhoroslov** [comment]
@ahmetmircik I did, pleasee have a look

**Comment by hasancelik** [comment]
run-lab-run

**Comment by hasancelik** [comment]
@dsukhoroslov pr-builder failed again due to checkstyle errors.

**Comment by dsukhoroslov** [comment]
yep, I see thouthands of checkstyle errors. But they do not relate to the changes that I did.

**Comment by dsukhoroslov** [comment]
@hasancelik found some errors in TopicConfig, fixed

**Comment by ahmetmircik** [comment]
run-lab-run

File: hazelcast/src/test/java/com/hazelcast/topic/TopicStressTest.java [hunk 4]
```diff
@@ -51,11 +52,16 @@
     private CountDownLatch startLatch;
     private PublishThread[] publishThreads;
     private Map<String, List<MessageListenerImpl>> listenerMap;
+    
+    @Parameterized.Parameter
```
Comment (jerrinot) [inline hazelcast/src/test/java/com/hazelcast/topic/TopicStressTest.java]: is parametrization working with the `HazelcastSerialClassRunner` ?

Comment (dsukhoroslov) [inline hazelcast/src/test/java/com/hazelcast/topic/TopicStressTest.java]: I have no idea. This change was requested by @ahmetmircik, see above.

Comment (dsukhoroslov) [inline hazelcast/src/test/java/com/hazelcast/topic/TopicStressTest.java]: Hi @jerrinot,

what the status of the PR, can it be merged?

Comment (jerrinot) [inline hazelcast/src/test/java/com/hazelcast/topic/TopicStressTest.java]: apparently the `HazelcastSerialClassRunner` supports parametrization thanks to #7720.

**Comment by jerrinot** [comment]
I think this PR is (almost) good to merge: The last thing is this: The property is now called `multi-threading`. It could be (mis-)interpreted as this promise: "When you set this property to `true` then Hazelcast will always use multiple threads to invoke the listener". 

I think something like `concurrent` might be more suitable. It's not a promise of what Hazelcast will do, it's more a capability of the listener: "This listener can be invoked concurrently". 

So for now there won't be any difference, but it reduces a change of misunderstanding. I can foresee a github issues saying: "I set the `multi-threading`, but Hazelcast is using only 1, (2, 3, 4...) threads to invoke my listener."

WDYT? Does it make sense? I think it's this is resolved then we can hit the merge button.

**Comment by dsukhoroslov** [comment]
I assume people will read javadoc at least. The javadoc states:
    /**
     * Enable multi-threaded message handling. When enabled any thread from events
     * thread pool can be used for incoming message processing. Otherwise only one
     * dedicated thread will be used to handle topic messages.
     * Note: it can be enabled only in case when global ordering is disabled. Moreover,
     * the local message ordering is not supported in this mode also. This means the
     * messages produced by local publisher can be processed by several threads with
     * no ordering guarantee.
     *
     * @param multiThreadingEnabled set to true to enable multi-threaded message processing, false to disable
     * @return The updated TopicConfig
     */

Do you find it confusing?

**Comment by jerrinot** [comment]
@dsukhoroslov: I see a subtle difference between:
1. Explicit configuration to use multiple threads. 
vs.
2. Configuration to indicate this listener has a *capability* to be invoked by concurrently

What do you think?

Anyway, even the current PR looks good. I just wanted to show it for completeness. If you think there is no real difference from a user-point of view then let me know. We can merge it as it is.

**Comment by dsukhoroslov** [comment]
Well, feature naming is a bit subjective thing, you know. Someone can find it confusing when others are not.. To be honest, I don't see the difference between `multiThreading` and `concurrent`. Let's use the same naming conventions as for other Topic properties: `globalOrdering`, `statistics`? I'd say they're more about _explicit configuration_ then a _capability_..

**Comment by jerrinot** [comment]
@dsukhoroslov: that's a good point with the other names. I'm going to merge this one. 

Many thanks for this contribution, it's very much appreciated!

**Comment by jerrinot** [comment]
@dsukhoroslov: and a very last thing: could you squash your commit into 1? There is no reason to have isolated commits related to checkstyle only. 

GitHub now has an ability to squash commits before merging them, but I want to make sure your name will stay in git history to give you the credit you surely deserve!

**Comment by dsukhoroslov** [comment]
Welcome, @jerrinot. Regarding the squash: it'll take some time. My present env is quite limited in connection, the `git fetch origin` throws error: `fatal: unable to access 'https://github.com/dsukhoroslov/hazelcast.git/': Failed to connect to github.com port 443: Connection refused`. Will try to do it from home in the evening.

**Comment by dsukhoroslov** [comment]
I did squash, please have a look. The last commit done by eclipse egit plugin after merge. Don't know how to avoid it.

**Comment by Donnerbart** [comment]
run-lab-run

**Comment by Donnerbart** [comment]
verify

**Comment by Donnerbart** [comment]
Git rebase is your friend. With that command you can pull the changes from another branch in yours and add your commit(s) on top of it again. But I don't think it's mandatory that you change this PR again.

**Comment by hasancelik** [comment]
run-lab-run

**Comment by Donnerbart** [comment]
run-lab-run

**Comment by jerrinot** [comment]
I merged this PR as extra commit does not overweight usefulness of this PR.
@dsukhoroslov: thank you for this contribution, it's much appreciated!

**Comment by dsukhoroslov** [comment]
Glad to hear this, thanks!

**Pull request merged**